Implementing Dynamic Observable Adapters for Embed Information fetching

alxjrvs opened the pull request.

alxjrvs committed:
Implementing QSDAdapter

alxjrvs committed:
implementing merge adapter, preparing for incoming FeedInfoAdapter

cqr reviewed (changes requested):
We need to make sure that the interface provided by the adapters is asynchronous (probably provided by an Rx Observable) and emits essentially the structure you described the adapters as having here.

cqr commented on src/app/embed/adapters/merge.adapter.ts:
```diff
@@ -0,0 +1,41 @@
+  }
+
+  get audioUrl(): string {
+    return this.QSDAdapter.audioUrl
```
You're constructing a new QSDAdapter for each of these properties.

alxjrvs committed:
Now uses single QSDAdapter, moving towards async

alxjrvs committed:
Initial stab at async MergeAdapter with sync QSDAdapter. 

will need to move Observable into QSDADapter with Advent of additional Adapters.

alxjrvs committed:
merging observables, possibly unconventionally.

alxjrvs committed:
save point

alxjrvs committed:
No longer a regression. Not yet complete.

cqr commented on src/app/embed/adapters/feed.adapter.ts:
```diff
@@ -0,0 +1,74 @@
+		let parser = new DOMParser();
+		let doc = <XMLDocument> parser.parseFromString(xml, 'application/xml');
+		let img: string;
+		let _img = Array.from(doc.querySelectorAll('channel > *[href]')).filter(e => e.nodeName === 'itunes:image')[0];
```
I think I would change this to `doc.querySelector(channel > *|image[href])`

cqr replied:
alternatively, if you're specifically looking for a namespace, there are non-querySelector methods that will let you pull those. Semantically, the thing that you're doing here is not quite right, though, because draper feeds don't include itunes:image, they include another, RP specific tag.

cqr commented on src/app/embed/adapters/feed.adapter.ts:
```diff
@@ -0,0 +1,74 @@
+		return (feedUrl || feedId)
+	}
+
+	private parseResponse(res: Response): Object {
```
I think you ought to define an interface instead of using the Object return type.

cqr commented on src/app/embed/adapters/merge.adapter.ts:
```diff
@@ -0,0 +1,38 @@
+import { Observable } from 'rxjs/Observable';
+
+export class MergeAdapter {
+  private QSDAdapter: QSDAdapter
```
Rather than automatically constructing and then keeping a set number of arguments, you should be accepting an array or varargs of "Adapter" arguments, where adapter is an interface that the different adapters implement. Probably as simple as`getParams(): Observable<AdapterProperties>` where `AdapterProperties` is the interface I described earlier.

cqr commented on src/app/embed/adapters/merge.adapter.ts:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  get getParams(): Observable<Object> {
+    if(this.QSDAdapter.hasFullInformation){
```
Would love some comments (either here on github or in code) to walk me through what's going on here.

cqr commented on src/app/embed/adapters/qsd.adapter.ts:
```diff
@@ -0,0 +1,58 @@
+    this.params = params
+  }
+
+  get hasFullInformation(): boolean {
```
I don't think this method is useful if you're thinking about this problem correctly.

cqr commented on src/app/embed/adapters/qsd.adapter.ts:
```diff
@@ -0,0 +1,58 @@
+    );
+  }
+
+  get getParams(): Observable<Object> {
```
Perfect. Again, probably a good idea to define an interface instead of using `Object` for the Observable type.

cqr commented on src/app/embed/embed.component.ts:
```diff
@@ -28,27 +28,33 @@ export class EmbedComponent implements OnInit {
+
+	ngOnInit() {
+		this.route.queryParams.forEach(params => {
+			const adapter = new MergeAdapter(params, this.http)
```
I think I would rather have this injected, rather than constructing it here. Also, the string property access is another indicator to me that we should define an interface.

cqr reviewed (changes requested):
a good step forward.

alxjrvs committed:
Save point

alxjrvs committed:
Base Adapter Structure created.

alxjrvs committed:
switched to AdapterProperties, moving towards true observable pattern

alxjrvs committed:
another stab at the same concept.

alxjrvs committed:
Observable Pattern now 1) merges all non-preferential adapters, then concats the results of the last one.

alxjrvs committed:
Services now Injected, no more instantiation (save for MergeAdapter). 

Pattern for selective observable injection being explored.

alxjrvs committed:
Commenting out unused privates

cqr commented on src/app/embed/adapters/adapter.properties.ts:
```diff
@@ -0,0 +1,34 @@
+export class  AdapterProperties {
```
I don't love the approach here. To me, this could be an interface (
```
+    public audioUrl?: string,
 +    public title?: string,
 +    public subtitle?: string,
 +    public subscribeUrl?: string,
 +    public subscribeTarget?: string,
 +    public artworkUrl?: string,
 +    public feedArtworkUrl?: string
```
) and a function. The notion of copying values from an object that literally matches the same interface is not great to me. LMK if you want to talk about how interfaces work in typescript.

alxjrvs replied:
set return value to AdapterProperties

cqr commented on src/app/embed/adapters/feed.adapter.ts:
```diff
@@ -0,0 +1,70 @@
+    })
+	}
+
+	private get draperUrl(): string { 
```
This could be using the same proxy url that the builder uses rather than pegging it directly to draper.

alxjrvs replied:
FEED_PROXY_URL

cqr commented on src/app/embed/adapters/feed.adapter.ts:
```diff
@@ -0,0 +1,70 @@
+        artworkUrl = artworkUrl[0].getAttribute('href')
+      };
+
+     let props = AdapterProperties.decode({ 
```
If this were an interface, you could just return `{ 
       audioUrl, 
       title, 
       subtitle, 
       subscribeUrl, 
       feedArtworkUrl,
       artworkUrl 
       }`

cqr commented on src/app/embed/adapters/merge.adapter.ts:
```diff
@@ -0,0 +1,31 @@
+import { EMBED_AUDIO_URL_PARAM, EMBED_TITLE_PARAM, EMBED_SUBTITLE_PARAM,
```
I'll need an overview of this. I think this adapter needs some love.

cqr commented on src/app/embed/adapters/qsd.adapter.ts:
```diff
@@ -0,0 +1,26 @@
+import { AdapterProperties } from './adapter.properties'
+
+@Injectable()
+export class QSDAdapter {
```
This is great!

cqr commented on src/app/embed/adapters/qsd.adapter.ts:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  private playerProperties(params): AdapterProperties {
+    return  AdapterProperties.decode({
```
again, could be just an interface :)

cqr commented on src/app/embed/embed.component.ts:
```diff
@@ -27,28 +29,49 @@ export class EmbedComponent implements OnInit {
+					this.subscribeUrl = ( properties.subscribeUrl || this.subscribeUrl )
+					this.subscribeTarget = ( properties.subscribeTarget || this.subscribeTarget ) 
+					this.artworkUrl = ( properties.artworkUrl || this.artworkUrl ) 
+          this.feedArtworkUrl = ( properties.feedArtworkUrl || this.feedArtworkUrl ) 
```
whitespace

cqr reviewed (changes requested):
We'll need to run through this again

alxjrvs committed:
class -> interface + func

alxjrvs committed:
Fixing up Observables to match requested patterns.

alxjrvs committed:
fix whitespace

cqr commented on src/app/embed/adapters/adapter.properties.ts:
```diff
@@ -0,0 +1,47 @@
+    (props.subscribeTarget !== undefined) &&
+    (props.artworkUrl !== undefined)
+}
+export function getMergedValues(...data: AdapterProperties[]):AdapterProperties {
```
It's confusing that this needs to be passed in reverse order (I know how we got here, but let's reverse here)

cqr commented on src/app/embed/adapters/feed.adapter.ts:
```diff
@@ -0,0 +1,68 @@
+
+      let subtitle = doc.getElementsByTagName('title')[0].textContent
+      let subscribeUrl = doc.getElementsByTagNameNS(atomNamespace, 'link')[0].getAttribute('href')
+      let feedArtworkUrl  = doc.getElementsByTagNameNS(rpNamespace, 'image')[0].getAttribute('href')
```
Please add a TODO comment in here to address the possibility of pulling non-rp-data in a future rev

cqr reviewed (changes requested):
Looking great!

alxjrvs committed:
now passing priority order to merging function, added comments to feed

cqr commented:
![](https://github.trello.services/images/mini-trello-icon.png) [QSD adapter for current query param structure](https://trello.com/c/tsdH2aBh/2-qsd-adapter-for-current-query-param-structure)

cqr commented:
![](https://github.trello.services/images/mini-trello-icon.png) [QSD adapter for current query param structure](https://trello.com/c/tsdH2aBh/2-qsd-adapter-for-current-query-param-structure)
